### PR TITLE
Add a @Timeout example for 'poll until' logic

### PR DESCRIPTION
**Pull request by rmannibucau**
## Overview

Proposing a pattern using @Timeout for "wait until" like tests.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

### CODE REVIEW

**Comment by sbrannen** [comment]
Tentatively slated for 5.5 GA for _team discussion_

**Comment by sbrannen** [comment]
I generally recommend the use of [Awaitility](https://github.com/awaitility/awaitility) for such use cases.

So, if we add an example like this in the User Guide we should also point the users to third-party solutions such as Awaitility.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 1]
```diff
@@ -1659,6 +1659,25 @@ omitted. Specifying no unit is equivalent to using seconds.
 | `42 d`          | `@Timeout(value = 42, unit = DAYS)`
 |===
 
+==== Use @Timeout for polling tests
+
+It is common to write tests waiting for some updates.
+In some cases you can rewrite the logic to use a `CountDownLatch` or any other synchronization mecanism but sometimes it is not possible - generally when depending on an external system.
+These tests require some timeout to ensure they don't hang the test suite forever and the test fails if the waiting condition "never" happens.
+With `@Timeout` it becomes very easy to write such tests since the test itself just require to become an infinite loop:
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Please format according to the project guidelines.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 2]
```diff
@@ -1659,6 +1659,25 @@ omitted. Specifying no unit is equivalent to using seconds.
...
+
+[source,java]
+----
+@Timeout(5) // 5s
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Please move the example to a java source file and ensure that it is executed in the build for the `documentation` project.

**Comment by sbrannen** [comment]
**Team Decision**: include such content, and we should also point the users to third-party solutions such as Awaitility.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 3]
```diff
@@ -1,5 +1,6 @@
 [[writing-tests]]
 == Writing Tests
+:testDir: /media/data/home/rmannibucau/1_dev/junit5/documentation/src/test/java
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Please remove that. `{testDir}` is already defined in `documentation.gradle.kts`.

Comment (rmannibucau) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: oops, thanks for the catch

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc (hunk 1 above)
Comment (AJNeufeld) [reply documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: "mecanism" should be "mechanism"

**Comment by sbrannen** [comment]
_in progress_: with the intent of fixing the Spotless violation and polishing locally before merging into `master`

**Comment by sbrannen** [comment]
This has been merged into `master` in e6006fe388dab0d8c4223b02f366fbe254516622 and further refined in 4f2da09aad0ddce47bdeb1c16dcb8661275d75a7.

Thanks for the PR!

In the future, please be mindful of the following.

- Example test classes in the `documentation` project must have the `Demo` suffix in order to be included in the build.
- Example test methods must be annotated with `@Test`.
- Sections in the User Guide require explicit anchors.

**Pull request closed without merging**